### DISCUSSION TITLE
Structured mesh creation/merge/combine

### DISCUSSION THREAD

**Discussion by ghost** [pyvista/pyvista#4972] [Q&A]
Hello,

I'm browsing python libraries on the internet for a project and recently found Pyvista which I don't know much about.

I have a cgns file that contains several blocks with a structured mesh, is it possible to use Pyvista to combine/merge/group all my blocks into a structured mesh?

My ultimate goal is to write my cgns file into a Tecplot or extract the mesh as a structured file.

Thanks for your help, have a nice day.
Toufik

**Comment by banesullivan** [comment]
If the meshes are truly structured, then yes -- you should be able to reconstruct and merge them with PyVista. Hard to tell without trying!

**Reply by ghost** [reply]
Thx for the reply :)
What do you mean by structured mesh (sorry to ask because apparently everyone has its on definition i am kind of lost) 

My CGNS file is a multi-blocks **structured** meshs (one can navigate inside each block using i, j and k indices). i exhausted all pyvista functions and only managed to retrieve an unstructured mesh. Do you have indications how i should tackle this with pyvista? i would like to integrate it in my simulation and pos-processing workflow

**Reply by banesullivan** [reply]
Apologies for my brevity, trying to respond to as many issues/discussions as I can

> What do you mean by structured mesh

I'm thinking they must conform to https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.StructuredGrid.html

https://docs.pyvista.org/version/stable/api/core/index.html#core-api-index

> A [pyvista.StructuredGrid](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.StructuredGrid.html#pyvista.StructuredGrid) is a regular lattice of points aligned with internal coordinate axes such that the connectivity can be defined by a grid ordering. These are commonly made from [numpy.meshgrid()](https://numpy.org/doc/stable/reference/generated/numpy.meshgrid.html#numpy.meshgrid). The cell types of structured grids must be 2D quads or 3D hexahedra. The [pyvista.StructuredGrid](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.StructuredGrid.html#pyvista.StructuredGrid) class is an extension of [vtk.vtkStructuredGrid](https://vtk.org/doc/nightly/html/classvtkStructuredGrid.html).

> only managed to retrieve an unstructured mesh. Do you have indications how i should tackle this with pyvista?

Hm. It sounds like the reader is giving instructed meshes. If there are structured, we may have to generate the new `StructuredGrid` mesh from this... hard help further  without having the data in hand


Though if all you care about is joining them, you can merge `UnstructuredGrid` meshes with: https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSetFilters.merge.html
